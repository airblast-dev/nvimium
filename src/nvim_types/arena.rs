use std::{
    cell::{RefCell, UnsafeCell},
    ffi::c_void,
    ptr::NonNull,
};

use libc::{c_char, c_double, size_t};
use thread_lock::call_check;

// arena_alloc_block for allocating
#[repr(C)]
#[derive(Clone, Debug)]
pub struct Arena {
    // TODO: make this public after adding cur block + c_char ptr
    cur_blk: Option<NonNull<c_char>>,
    pub pos: size_t,
    pub size: size_t,
}

thread_local! {
    pub static CALLBACK_ARENA: RefCell<Arena> = const { RefCell::new(Arena::EMPTY) };
}
const _: () = assert!(size_of::<Arena>() == size_of::<*mut c_char>() + size_of::<size_t>() * 2);

impl Arena {
    pub const EMPTY: Arena = Arena {
        cur_blk: None,
        pos: 0,
        size: 0,
    };

    /// Reset the position of the arena
    ///
    /// Unlike setting the `pos` field to 0 this will minimally keep enough space to store the
    /// previous block information. In case you are using an arena and want to reclaim space
    /// without dealing with neovim functions use this.
    ///
    /// While this function is safe, incorrect usage of the Arena after calling this function can
    /// cause UB.
    pub fn reset_pos(&mut self) {
        self.pos = arena_align_offset(size_of::<ArenaMem>() as u64);
    }
}

impl Drop for Arena {
    fn drop(&mut self) {
        // SAFETY: we are mutating a neovim static in arena_finish, we must have execution yielded
        // in order to drop this
        if self.cur_blk.is_some() {
            call_check();
            unsafe { arena_mem_free(arena_finish(self)) };
        }
    }
}

pub type ArenaMem = UnsafeCell<*mut ConsumedBlk>;

#[derive(Clone, Debug)]
#[repr(C)]
pub struct ConsumedBlk {
    prev: *mut Self,
}

/// The minimum space needed to store the prev block ptr.
fn arena_align_offset(off: u64) -> size_t {
    // no idea why we do this but just we are just replicating whats done in the same function in
    // neovim
    const ARENA_ALIGN: usize = {
        if size_of::<c_double>() >= size_of::<*mut c_void>() {
            size_of::<c_double>()
        } else {
            size_of::<*mut c_void>()
        }
    };

    // instead of the size of ArenaMem we might be able to find a way to recover these precious
    // bytes where a pointer isn't stored in it
    //
    // sadly difficult to do as we have no way to know how the allocation was requested in neovim
    // (namely the align value passed to allocator functions)
    //
    // worst case we unnecesarily use the size of a pointer
    (off as usize + (ARENA_ALIGN - 1)) & !(ARENA_ALIGN - 1)
}

unsafe extern "C" {
    pub(crate) fn arena_mem_free(arena_mem: ArenaMem);
    pub(crate) fn arena_finish(arena: *mut Arena) -> ArenaMem;
}
