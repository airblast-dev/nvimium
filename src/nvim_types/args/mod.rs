pub mod open_term_cb;
