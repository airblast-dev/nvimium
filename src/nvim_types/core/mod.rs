pub mod array;
pub(crate) mod borrowed;
pub mod buffer;
pub mod call_site;
pub mod dictionary;
pub mod error;
pub mod hl_group;
pub mod kvec;
pub mod lua_ref;
pub mod namespace;
pub mod object;
pub mod string;
pub mod tab_page;
pub mod window;

pub type Integer = i64;
pub type Boolean = bool;
pub type Float = libc::c_double;
pub type HandleT = libc::c_int;
pub type HLGroupIDT = Integer;
pub type LuaRefT = libc::c_int;

pub use array::Array;
pub use buffer::Buffer;
pub use call_site::Channel;
pub use dictionary::{Dictionary, KeyValuePair};
pub use error::Error;
pub use hl_group::HlGroupId;
pub use kvec::KVec;
pub use lua_ref::LuaRef;
pub use namespace::NameSpace;
pub use object::Object;
pub use string::{AsThinString, OwnedThinString, String, ThinString};
pub use tab_page::TabPage;
pub use window::Window;
