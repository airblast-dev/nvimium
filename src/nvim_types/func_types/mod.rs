pub mod create_user_command;
pub mod echo;
pub mod feedkeys;
pub mod keymap_mode;
