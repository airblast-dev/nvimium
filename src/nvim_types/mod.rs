extern crate alloc;

mod nvalloc;

pub mod args;
pub mod arena;
pub mod core;
pub mod func_types;
pub mod lua;
pub mod object_subs;
pub mod opts;
pub mod returns;

pub use arena::*;
pub use core::*;
pub use lua::{FromLua, IntoLua};
