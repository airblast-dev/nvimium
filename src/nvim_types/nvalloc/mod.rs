// enables error recovery for allocating functions
#[cfg(not(any(miri, test)))]
mod nvdefs;

use std::{ffi::c_void, ptr::NonNull};

use libc::{c_char, malloc, size_t};

#[cfg(not(any(miri, test)))]
pub(crate) use nvdefs::{E_OUTOFMEM, preserve_exit, try_to_free_memory};
use panics::alloc_failed;

#[cfg(not(any(miri, test)))]
use thread_lock::can_call;

#[inline]
pub(crate) fn real_size(size: size_t, count: size_t) -> size_t {
    size.checked_mul(count).unwrap()
}

pub unsafe fn xmalloc(size: size_t, count: size_t) -> NonNull<c_void> {
    let real_size = real_size(size, count);

    if real_size == 0 {
        #[cold]
        #[inline(never)]
        fn cold() {}
        cold();
        return NonNull::dangling();
    }
    #[allow(unused_mut)]
    let mut ptr = unsafe { malloc(real_size) };
    #[cfg(not(any(miri, test)))]
    if ptr.is_null() && can_call() {
        unsafe {
            try_to_free_memory();
            ptr = malloc(real_size);
        }
    }

    if ptr.is_null() {
        alloc_failed();
    }

    NonNull::new(ptr).unwrap()
}

pub unsafe fn xrealloc(
    ptr: *mut c_void,
    size: size_t,
    old_cap: size_t,
    new_cap: size_t,
) -> NonNull<c_void> {
    let real_size = real_size(size, new_cap);

    if real_size == 0 {
        if old_cap != 0 {
            unsafe {
                libc::free(ptr);
            }
        }
        return NonNull::dangling();
    }

    let ptr = unsafe {
        #[allow(unused_mut)]
        let mut new_ptr = libc::realloc(ptr, real_size);

        // we couldnt allocate memory and execution is yielded to us
        // tell neovim to free up some memory and try allocating again
        #[cfg(not(any(miri, test)))]
        if new_ptr.is_null() && can_call() {
            {
                try_to_free_memory();
                new_ptr = libc::realloc(ptr, real_size);
            }

            // we already tried to recover some memory but failed, preserve all files and whatever else
            // the function does so there is no data loss
            if new_ptr.is_null() {
                preserve_exit(E_OUTOFMEM);
            }
        }

        if new_ptr.is_null() {
            alloc_failed()
        }

        new_ptr
    };

    NonNull::new(ptr).unwrap()
}

pub unsafe fn xfree(ptr: *mut c_void, capacity: size_t) {
    if capacity != 0 {
        unsafe {
            libc::free(ptr);
        }
    }
}

#[inline]
#[must_use]
pub unsafe fn xmemdupz(src: *const c_void, len: usize, size: size_t) -> NonNull<c_void> {
    unsafe {
        let ptr = xmalloc(size, len.checked_add(1).unwrap());
        xmemcpyz(ptr.as_ptr(), src, len, size);
        ptr
    }
}

#[inline]
pub unsafe fn xmemcpyz(dst: *mut c_void, src: *const c_void, len: usize, size: size_t) {
    let byte_len = len.checked_mul(size).unwrap();
    assert!(!dst.is_null() && !src.is_null());
    unsafe {
        libc::memcpy(dst, src, byte_len);
        *dst.byte_add(byte_len).cast::<c_char>() = 0;
    }
}
