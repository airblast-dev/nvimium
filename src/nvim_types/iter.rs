use core::slice;
use std::iter::{FusedIterator, Map};

use super::{Object, ThinString};

pub struct ThIter<'a> {
    inner: Map<slice::Iter<'a, Object>, fn(&'a Object) -> ThinString<'a>>,
}

impl<'a> ThIter<'a> {
    pub(crate) fn new(s: &'a [Object]) -> Self {
        Self {
            inner: s.iter().map(|obj| obj.as_string().unwrap().as_thinstr()),
        }
    }
}

impl<'a> Iterator for ThIter<'a> {
    type Item = ThinString<'a>;
    fn next(&mut self) -> Option<Self::Item> {
        self.inner.next()
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        self.inner.size_hint()
    }

    fn count(self) -> usize
    where
        Self: Sized,
    {
        self.inner.count()
    }

    fn last(self) -> Option<Self::Item>
    where
        Self: Sized,
    {
        self.inner.last()
    }

    fn nth(&mut self, n: usize) -> Option<Self::Item> {
        self.inner.nth(n)
    }
}

impl<'a> DoubleEndedIterator for ThIter<'a> {
    fn next_back(&mut self) -> Option<Self::Item> {
        self.inner.next_back()
    }

    fn nth_back(&mut self, n: usize) -> Option<Self::Item> {
        self.inner.nth_back(n)
    }
}

impl<'a> ExactSizeIterator for ThIter<'a> {}
impl<'a> FusedIterator for ThIter<'a> {}
