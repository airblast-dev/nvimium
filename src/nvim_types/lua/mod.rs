// Using mlua instead of mlua-sys would remove the need of many of the things in this module
//
// Problem is the function struct it provides works via the lua stack and not with the lua
// registry. This means we cannot pass a function to neovim via mlua.
//
// Instead this is a partial reimplimentation of the functionality of mlua but aims to only support
// callbacks passed to neovim.
//
// The goal here is to provide the minimum requirements to be able to effectively use the C api.
// If lua functions are intended to be called directly users should use mlua in combination with
// nvimium. In some cases nvimium may provide convenience traits to simplify things but this is not
// a guarantee as some things in neovim dont map well to mlua's API (mlua is really great, just not
// suitable here).
//
// while some parts could still be used from mlua it is not really worth bringing in another dependency
// as all use cases have a fix set of arguments which we can handle internally
mod box_fn;
mod closure;
pub mod core;
mod fn_ptr;
#[doc(hidden)]
pub use box_fn::set_callback_name;

use core::FromLuaMany;
pub use core::{FromLua, IntoLua};
use std::any::Any;

#[doc(hidden)]
pub use mlua_sys::lua_State;
use thread_lock::get_lua_ptr;

use crate::nvim_types::LuaRef;

#[cfg(target_pointer_width = "32")]
type LuaInteger = i32;
#[cfg(target_pointer_width = "64")]
type LuaInteger = i64;

#[repr(transparent)]
pub struct Function(LuaRef);

impl Function {
    pub(crate) fn from_fn<F: 'static + Fn(A) -> R + Unpin, A: FromLua, R: IntoLua>(f: F) -> Self {
        let mut l = get_lua_ptr();
        Self(unsafe { LuaRef::new(closure::register(l.as_ptr(), f)) })
    }

    pub(crate) fn from_fn_ptr<A: FromLuaMany, R: IntoLua>(f: fn(A) -> R) -> Self {
        let mut l = get_lua_ptr();
        Self(unsafe { LuaRef::new(fn_ptr::register(l.as_ptr(), f)) })
    }

    pub(crate) fn from_box_fn<F: 'static + Fn(A) -> R, A: FromLuaMany, R: IntoLua>(f: F) -> Self {
        let mut l = get_lua_ptr();
        Self(unsafe { LuaRef::new(box_fn::register(l.as_ptr(), f)) })
    }

    pub fn into_luaref(self) -> LuaRef {
        self.0
    }

    /// Wraps the provided function and returns a [`Function`] that stores the key to the lua registry
    ///
    /// The provided function will be attempted to be downcasted to a function pointer for cheaper
    /// initialization and drops.
    pub fn wrap<F: 'static + Fn(A) -> R + Unpin, A: 'static + FromLuaMany, R: 'static + IntoLua>(
        f: F,
    ) -> Self {
        // if F is a function pointer we can avoid dynamic dispatch, an extra indirection and the drop
        // call passed to lua (we can use lightuserdata)
        if let Some(f) = (&f as &dyn Any).downcast_ref::<fn(A) -> R>() {
            Self::from_fn_ptr(*f)
        } else {
            Self::from_box_fn(f)
        }
    }
}
