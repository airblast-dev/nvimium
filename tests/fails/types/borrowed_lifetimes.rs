use std::ops::DerefMut;

use nvim_types::{array::Array, borrowed::Borrowed};

fn main() {
    fn from_array() {
        let array = Array::default();
        let borrowed: Borrowed<'_, Array> = array.into();
    }

    fn from_array_mut() {
        let array = Array::default();
        let borrowed: Borrowed<'_, Array> = (&array).into();
        let mut_array = array.deref_mut();
        borrowed.as_ref();
    }
}
