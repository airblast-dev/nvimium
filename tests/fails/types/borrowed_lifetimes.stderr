error[E0433]: failed to resolve: use of unresolved module or unlinked crate `nvim_types`
 --> tests/fails/types/borrowed_lifetimes.rs:3:5
  |
3 | use nvim_types::{array::Array, borrowed::Borrowed};
  |     ^^^^^^^^^^ use of unresolved module or unlinked crate `nvim_types`
  |
help: there is a crate or module with a similar name
  |
3 - use nvim_types::{array::Array, borrowed::Borrowed};
3 + use nvim_test::{array::Array, borrowed::Borrowed};
  |
